Add DpWriter

bocchino opened the pull request:
This PR adds the `Svc/DpWriter` component and supporting unit tests. It also does the following:
* Add data product-specific configuration files to `config`.
* Add a `Fw::FileNameString` object for storing file names. The buffer size is configurable in FPP.
* Add more fine-grained control of the error status return values in the unit test Os::File stub.
* Add a method to `Utils::HashBuffer` for converting the first four bytes of the hash to a `U32` value. This is useful for error reporting.
* Do incidental reformatting of `Utils/HashBuffer.{hpp,cpp}`.
* Add data product-specific ports to `Svc/DpPorts`.
* Make a few incidental changes to `Svc/DpManager`.
* Fix spelling of header files included from `config`, as discussed in #2640.

bocchino committed:
Pull in framework changes from data-products branch

bocchino committed:
Pull in changes to DpManager from data-products branch

bocchino committed:
Pull in DpWriter from data-products branch

bocchino committed:
Fix spelling

bocchino committed:
Revise FileNameString

bocchino commented:
Something is wrong with the dp-writer branch on GitHub. It's not picking up my changes.

bocchino committed:
Fix warnings in CI

bocchino committed:
Fix static analysis warnings

bocchino commented on Utils/Hash/HashBufferCommon.cpp:
```diff
@@ -1,55 +1,70 @@
-        Fw::SerializeStatus stat = Fw::SerializeBufferBase::setBuff(other.m_bufferData,other.getBuffLength());
-        FW_ASSERT(Fw::FW_SERIALIZE_OK == stat,static_cast<NATIVE_INT_TYPE>(stat));
-    }
+HashBuffer::HashBuffer(const U8* args, NATIVE_UINT_TYPE size) {
```
Fixed

bocchino committed:
Fix static analysis warnings

bocchino committed:
Revise formatting and comments

bocchino committed:
Revise banner comments

timcanham commented on Fw/Types/FileNameString.hpp:
```diff
@@ -0,0 +1,36 @@
+    NATIVE_UINT_TYPE getCapacity() const;  //!< return buffer size
+
+  private:
+    char m_buf[STRING_SIZE];  //!< storage for string data
```
Could we make `STRING_SIZE` a separate config variable like `FILE_STRING_SIZE`?

bocchino replied:
`FileNameString::STRING_SIZE` is an alias for `FileNameStringSize`, which is defined in `config/AcConstants.fpp`.

timcanham replied:
If you are making any changes, perhaps `FileNameString::STRING_SIZE` could be the array size to make the scope explicit?

bocchino replied:
Sure, I can add that qualifier.

bocchino replied:
Done.

timcanham commented on Fw/Types/FileNameString.cpp:
```diff
@@ -0,0 +1,54 @@
+        return *this;
+    }
+
+    (void)Fw::StringUtils::string_copy(this->m_buf, other.toChar(), sizeof(this->m_buf));
```
Not sure if this is a general approach for strings, but should we use [length](https://github.com/nasa/fprime/blob/devel/Fw/Types/StringType.hpp#L28) of source string to avoid overcopying?

bocchino replied:
We want to use the size of the destination buffer, to avoid overrunning the buffer. For example, a string of size 100 that's copied into a buffer of size 40 will be truncated.

timcanham replied:
You're right on that part that we want to limit the max to be the local buffer size. How about a minimum of the source actual string size and the local storage size?

bocchino replied:
The string copy utility does do a minimum. It copies the actual length of the source string unless that length is greater than the buffer size, in which case it truncates the string to fit in the buffer size. This is a standard size-bounded copy operation. It's just like strncpy, except that it's safer.

timcanham commented on Svc/DpPorts/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+####
```
Why wouldn't these go in `Fw/Dp`?

bocchino replied:
I followed the pattern of ports like `Svc::Cycle` that are used only in the Svc layer.

timcanham replied:
The difference is that `Svc::Cycle` isn't in the FP spec, where DP things are. I think we want a pattern of FPP supported types being in `Fw`. Thoughts @LeStarch?

bocchino replied:
These ports aren't in the FPP spec either. They're in the F Prime data products design. I think Cycle is similar. It supports the Svc F Prime rate group components and their uses; these ports support the Svc data product components and their uses.

bocchino replied:
The only port that's currently in Svc/DpPorts is DpWritten, which is the port that DpWriter uses to communicate to DpCatalog. I think it belongs in Svc. You're right that the FPP special ports (DpGet, etc.) belong in Fw/Dp. That is where they are.

timcanham reviewed (changes requested):
Just a few comments...

bocchino committed:
Revise FileNameString per PR comment

LeStarch commented:
@bocchino @timcanham ready for merge?

timcanham commented:
![image](https://github.com/nasa/fprime/assets/8086374/bb4bfdb3-9bef-438c-9232-fe3f51191f1c)

bocchino committed:
Revise path names in config headers

If a header H.hpp exists in the F Prime source base, then

`#include "config/H.hpp"`

is dangerous. Because [project root] and [fprime root] are both
in the list of include paths, it's not clear whether this means "include
[project root]/config/H.hpp" or "include [fprime root]/config/H.hpp."

On the other hand,

include <config/H.hpp>

or

`#include "config/H.hpp"`

has no such ambiguity, because only one of [project root]/config
and [fprime root]/config is in the list of include paths.

bocchino committed:
Revise path names in config headers

If a header H.hpp exists in the F Prime source base, then

is dangerous. Because [project root] and [fprime root] are both
in the list of include paths, it's not clear whether this means "include
[project root]/config/H.hpp" or "include [fprime root]/config/H.hpp."

On the other hand,

or

has no such ambiguity, because only one of [project root]/config
and [fprime root]/config is in the list of include paths.

bocchino committed:
Merge branch 'dp-writer' of github.com:bocchino/fprime into dp-writer

bocchino commented:
Ready! I made some changes to config include paths as discussed.

LeStarch commented:
Wooooooooooooo...... :shipit:
